README.md

<div id="top"></div>

<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![LinkedIn][linkedin-shield]][linkedin-url]
-->

[![MIT License][license-shield]][license-url]

<!-- PROJECT LOGO -->
<div align="center">
  <a href="https://github.com/othneildrew/Best-README-Template"><img src="images/logo.jfif" alt="Logo"></a>
  <h3 align="center">Rohde&Schwarz HMP2020 Series Remote Control with Go</h3>
  <p align="center"> Use USB Virtual Com Port for switching the Power Supply</p>
</div>

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>

<!-- vscode-markdown-toc -->

- [1. About The Project](#1-about-the-project)
- [2. Usage](#2-usage)
- [3. License](#3-license)

<!-- vscode-markdown-toc-config
	numbering=true
	autoSave=true
	/vscode-markdown-toc-config -->
<!-- /vscode-markdown-toc -->

  </ol>
</details>

<!-- ABOUT THE PROJECT -->
##  1. <a name='AboutTheProject'></a>About The Project

This package provides a Go package you can use for HMP2020 remote control.

<p align="right">(<a href="#top">back to top</a>)</p>

##  2. <a name='Usage'></a>Usage

You can use the demo app, the **hmp** executable, in this way without building the software to generate power cycles for testing:

```bash
hmp -p com7 -ch 2 -V 24 -mA 1000 -msON 5000 -msOFF 1000 -v
```

* Uses COM port 7 (as example)
* Control channel 2
* Set channel to 24 Volt output voltage.
* Set channel to 1000 mA max current.
* Start endless loop:
  * Switch channel ON for 5 seconds.
  * Switch channel OFF for 1 second.
* Show all commands and measured values.

<!-- LICENSE -->
##  3. <a name='License'></a>License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/othneildrew/Best-README-Template.svg?style=for-the-badge
[contributors-url]: https://github.com/othneildrew/Best-README-Template/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/othneildrew/Best-README-Template.svg?style=for-the-badge
[forks-url]: https://github.com/othneildrew/Best-README-Template/network/members
[stars-shield]: https://img.shields.io/github/stars/othneildrew/Best-README-Template.svg?style=for-the-badge
[stars-url]: https://github.com/othneildrew/Best-README-Template/stargazers
[issues-shield]: https://img.shields.io/github/issues/othneildrew/Best-README-Template.svg?style=for-the-badge
[issues-url]: https://github.com/othneildrew/Best-README-Template/issues
[license-shield]: https://img.shields.io/github/license/othneildrew/Best-README-Template.svg?style=for-the-badge
[license-url]: https://github.com/othneildrew/Best-README-Template/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/othneildrew
[product-screenshot]: images/screenshot.png


CHANGELOG.md
# Changelog

## v0.0.0 Initial Release (untested)

## v0.1.0 Functional Release (tested)

### Added or Changed

* The `hmp` package is usable.
* Working executable as example for the `hmp` package usage added.

## v0.2.0 Refactored Release (tested)

* The internal interface is improved.

## v0.3.0 Refactored Release (tested)

* The internal interface is improved.

## v0.3.1 Refactored Release (tested)

* This is a minor improvement.

## v0.3.2 Refactored Release (tested)

* This is a minor correction.

## v0.3.3 Refactored Release (tested)

* This is a minor improvement.
  
## v0.3.4 Refactored Release (tested)

* Comport label added.

## v0.3.5 Refactored Release (tested)

* This is a minor correction.

## v0.3.6 Refactored Release (tested)

* This is a minor correction.

## v0.4.0 CLI switches added

* Command line interface switches can control the HMP device.


pkg/com/com.go
// Copyright 2020 Thomas.Hoehenleitner [at] seerose.net
// Use of this source code is governed by a license that can be found in the LICENSE file.

// Package com reads from COM port.
package com

import (
	"fmt"
	"io"
	"log"
	"os"
	"strings"

	"go.bug.st/serial"
)

// COMport is the comport interface type to use different COMports.
type COMport interface {
	Open() bool
	Read(buf []byte) (int, error)
	Write(buf []byte) (int, error)
	Close() error
}

// Port is a serial device trice receiver
type Port struct {
	verbose      bool
	port         string
	serialHandle serial.Port
	serialMode   serial.Mode
	w            io.Writer
	label        string
}

// NewPort creates an instance of a serial device type trice receiver
func NewPort(w io.Writer, label string, comPortName string, baud int, dataBits int, parity string, stopBits string, verbose bool) *Port {
	var parityB serial.Parity
	switch strings.ToLower(parity) {
	case "n", "no", "none":
		parityB = serial.NoParity
	case "e", "ev", "even":
		parityB = serial.EvenParity
	case "o", "odd":
		parityB = serial.OddParity
	default:
		log.Fatal("invalid parity value: ", parity, " Accepting case insensitive: n|no|none|e|even||o|odd.")
	}

	var stopB serial.StopBits
	switch strings.ToLower(stopBits) {
	case "1", "one":
		stopB = serial.OneStopBit
	case "1.5":
		stopB = serial.OnePointFiveStopBits
	case "2", "two":
		stopB = serial.TwoStopBits
	default:
		log.Fatalf(" Unknown stop bits value \"%s\". Valid are \"1\", \"1.5\". \"2\"\n", stopBits)
	}

	if !(5 <= dataBits && dataBits <= 9) {
		log.Fatalf("Invalid dataBits value %d. Valid are 5-9\n", dataBits)
	}

	r := &Port{
		port: comPortName,
		serialMode: serial.Mode{
			BaudRate: baud,
			DataBits: dataBits,
			Parity:   parityB,
			StopBits: stopB,
		},
	}

	r.label = label
	r.w = w
	r.verbose = verbose
	if verbose {
		fmt.Fprintln(w, "New COM port:", r)
	}
	return r
}

// Read blocks until (at least) one byte is received from
// the serial port or an error occurs.
// It stores data received from the serial port into the provided byte array
// buffer. The function returns the number of bytes read.
func (p *Port) Read(buf []byte) (int, error) {
	return p.serialHandle.Read(buf)
}

func (p *Port) Write(buf []byte) (int, error) {
	return p.serialHandle.Write(buf)
}

// Close releases port.
func (p *Port) Close() error {
	if p.verbose {
		fmt.Fprintln(p.w, "Closing "+p.label+" COM port.")
	}
	return p.serialHandle.Close()
}

// Open initializes the serial receiver.
//
// It opens a serial port and returns true on success.
func (p *Port) Open() bool {
	var err error
	p.serialHandle, err = serial.Open(p.port, &p.serialMode)
	if err != nil {
		if p.verbose {
			fmt.Fprintln(p.w, err, "try '", os.Args[0], "s' to check for serial ports")
		}
		return false
	}
	return true
}

// GetSerialPorts scans for serial ports.
func GetSerialPorts(w io.Writer, verbose bool) ([]string, error) {
	ports, err := serial.GetPortsList()

	if err != nil {
		fmt.Fprintln(w, err)
		return ports, err
	}
	if len(ports) == 0 {
		if verbose {
			fmt.Fprintln(w, "No serial ports found!")
		}
		return ports, err
	}
	for _, port := range ports {
		pS := NewPort(w, "try", port, 115200, 8, "N", "1", false)
		if pS.Open() {
			pS.Close()
			fmt.Fprintln(w, "Found port: ", port)
		} else {
			fmt.Fprintln(w, "Found port: ", port, "(used)")
		}
	}
	return ports, err
}


pkg/hmp/hmp.go
package hmp

import (
	"errors"
	"fmt"
	"io"
	"log"
	"strings"
	"time"

	"github.com/rokath/HMP2020-Go-Bindings/pkg/com"
)

var (

	// SerialPortName is the OS specific serial port name used for HMP device.
	SerialPortName string

	// Parity is the transmitted bit parity: "even", "odd", "none"
	Parity string

	// BaudRate is the configured baud rate of the serial port. It is set as command line parameter.
	BaudRate int

	// DataBits is the serial port bit count for one "byte".
	DataBits int

	// StopBits is the number of stop bits: "1", "1.5", "2"
	StopBits = "1"

	// Power is the power device to be controlled.
	Power *Device
)

type Device struct {

	// Ports is the assigned communictation port.
	Port *com.Port

	verbose bool

	w io.Writer
}

func NewDevice(w io.Writer, serialPortName string, baudRate int, dataBits int, parity string, stopBits string, verbose bool) (p *Device) {
	p = new(Device)
	p.verbose = verbose
	p.w = w
	p.Port = com.NewPort(w, "HMP", SerialPortName, baudRate, dataBits, parity, stopBits, verbose)

	if !p.Port.Open() {
		log.Fatal(errors.New(serialPortName + " port failure, try with -v for more information"))
	}
	return
}

// Read blocks until (at least) one byte is received from
// the serial port or an error occurs.
// It stores data received from the serial port into the provided byte array
// buffer. The function returns the number of bytes read.
func (p *Device) Read(buf []byte) (int, error) {
	return p.Port.Read(buf)
}

func (p *Device) Write(buf []byte) (int, error) {
	return p.Port.Write(buf)
}

// Close releases port.
func (p *Device) Close() error {
	return p.Port.Close()
}

// Connect tries to get contact to HMP2020.
func (p *Device) Connect() error {
	response, e := p.SendAndReceive("*IDN?\n", 100) // 50 is the fractal border.
	if e != nil {
		return e
	}
	exp := "ROHDE&SCHWARZ,HMP2020"
	if response[:len(exp)] != exp {
		return fmt.Errorf("HMP2020 not connected")
	}
	if p.verbose {
		fmt.Println(strings.TrimSuffix(response, "\n"), "connected")
	}
	return nil
}

// Send transmits cmd and waits ms afterwards before returning.
func (p *Device) Send(cmd string, ms time.Duration) {
	b := []byte(cmd)
	m, e := p.Write(b)
	if m != len(b) || e != nil {
		fmt.Printf("Wrote only %d bytes and not %d bytes, error %v", m, len(b), e)
	}
	time.Sleep(ms * time.Millisecond)
}

// SendAndReceive transmits cmd, waits ms milliseconds and returns response in r.
// The returned string is "as is" from HMP2020.
func (p *Device) SendAndReceive(cmd string, ms time.Duration) (r string, e error) {
	p.Send(cmd, ms)       // needs a line end
	b := make([]byte, 64) // 32 needed
	var n int
	n, e = p.Read(b)
	if e != nil {
		fmt.Println(e)
	}
	if n == 0 {
		e = errors.New("no answer from HMP2020")
		return "", e
	}
	b = b[:n]
	r = string(b)
	return r, e
}

func (p *Device) Query(cmd string) (response string) {
	if p.verbose {
		fmt.Println("query:", cmd)
	}
	response, e := p.SendAndReceive(cmd+"\r\n", 500) // 50 is the fractal border.
	if e != nil {
		fmt.Println(e) // log.Fatal(e)
	}
	return
}

func (p *Device) Command(cmd string) {
	if p.verbose {
		fmt.Println("command:", cmd)
	}
	p.Send(cmd+"\r\n", 100) // 50 is the fractal border.
}


pkg/hmp/commands.go
package hmp

import (
	"strconv"
)

// OutputOFF disables output on channel n. Valid channel numbers are 1, 2, 3, 4.
// n=0 or -1 switches all channel outputs.
func (p *Device) OutputOFF(n int) {
	if n <= 0 {
		p.Command("OUTPUT:GENERAL OFF")
		return
	}
	ch := strconv.Itoa(n)
	p.Command("INST:SEL OUT" + ch)
	p.Command("OUTPUT:SELECT OFF")
}

// OutputON enables output on channel n. Valid channel numbers are 1, 2, 3, 4.
// n=0 or -1 switches all channel outputs.
func (p *Device) OutputON(n int) {
	if n <= 0 {
		p.Command("OUTPUT:GENERAL ON")
		return
	}
	ch := strconv.Itoa(n)
	p.Command("INST:SEL OUT" + ch)
	p.Command("OUTPUT:SELECT ON")
}

// SetVoltage sets voltage on channel n. Valid channel numbers are 1, 2, 3, 4.
func (p *Device) SetVoltage(n int, v string) {
	ch := strconv.Itoa(n)
	p.Command("INST:SEL OUT" + ch)
	p.Command("SOURCE:VOLTAGE:LEVEL " + v)
}

// Voltage returns voltage on channel n. Valid channel numbers are 1, 2, 3, 4.
func (p *Device) Voltage(n int) (v string) {
	ch := strconv.Itoa(n)
	p.Command("INST:SEL OUT" + ch)
	return p.Query("MEASURE:SCALAR:VOLTAGE:DC?")
}

// SetCurrent sets max current on channel n. Valid channel numbers are 1, 2, 3, 4.
func (p *Device) SetCurrent(n int, v string) {
	ch := strconv.Itoa(n)
	p.Command("INST:SEL OUT" + ch)
	p.Command("SOURCE:CURRENT:LEVEL " + v)
}

// Current returns current on channel n. Valid channel numbers are 1, 2, 3, 4.
func (p *Device) Current(n int) (v string) {
	ch := strconv.Itoa(n)
	p.Command("INST:SEL OUT" + ch)
	return p.Query("MEASURE:SCALAR:CURRENT:DC?")
}


internal/args/handler.go
package args

import (
	"flag"
	"fmt"
	"io"
	"log"
	"time"

	"github.com/rokath/HMP2020-Go-Bindings/pkg/com"
	"github.com/rokath/HMP2020-Go-Bindings/pkg/hmp"
	"github.com/spf13/afero"
)

var (
	// Version is the program version number and is injected from main package.
	Version string

	// Commit is the program checksum and injected from main package.
	Commit string

	// Date is the compile time and injected from main package.
	Date string

	// Verbose, if true gives additional information for issue fixing.
	verbose bool

	// Version is the program version number and is injected from main package.
	VersionFlag bool

	// scanComPorts is used to find com ports.
	scanComPorts bool

	// channel is the output channel.
	channel int

	// voltage is the channel specific voltage.
	voltage string

	// current is the channel specific max current in mA
	current string

	// timeOn is the channel specific ON time in ms.
	timeOn int

	// timeOff is the channel specific OFF time in ms.
	timeOff int
)

func init() {
	flag.BoolVar(&scanComPorts, "s", false, "Scan com ports")
	flag.BoolVar(&VersionFlag, "version", false, "Show version information.")
	flag.BoolVar(&verbose, "v", false, "Show verbose messages")
}

func init() {
	flag.StringVar(&hmp.SerialPortName, "p", "", "Short for hmpPort")
	flag.StringVar(&hmp.SerialPortName, "hmpPort", "", "Use Port for HMP2020 or HMP4040")
	flag.IntVar(&hmp.BaudRate, "hmpBaud", 9600, "Set HMP baud rate.")
	flag.IntVar(&hmp.DataBits, "hmpDataBits", 8, "Set HMP data bit count.")
	flag.StringVar(&hmp.Parity, "hmpParity", "none", "Set parity")
	flag.StringVar(&hmp.StopBits, "hmpStopBits", "1", "Set parity")
	flag.IntVar(&channel, "hmpChannel", 1, "Select HMP channel.")
	flag.IntVar(&channel, "ch", 1, "Short for hmpChannel.")
}

func init() {
	flag.StringVar(&voltage, "V", "1.7", "Set channel output voltage.")
	flag.StringVar(&current, "mA", "10", "Set channel max current.")
	flag.IntVar(&timeOn, "msON", 5000, "Set channel time ON.")
	flag.IntVar(&timeOff, "msOFF", 1000, "Set channel time OFF.")

}

// Handler is called in main, evaluates args and calls the appropriate functions.
// It returns for program exit.
// All output is written to w.
// fSys is used aas file system.
func Handler(w io.Writer, fSys *afero.Afero, args []string) error {

	if Date == "" { // goreleaser will set Date, otherwise use file info.
		fi, err := fSys.Stat(args[0])
		if nil == err { // On running main tests file-info is invalid, so do not use in that case.
			Date = fi.ModTime().String()
		}
	}

	flag.Parse()

	if flag.NFlag() == 0 { // no CLI flags
		fmt.Println(`Enter "hmp -help"`)
		return nil
	}

	if VersionFlag {
		if Version != "" {
			fmt.Print("version=", Version)
		}
		if Commit != "" {
			fmt.Print("commit=", Commit)
		}
		fmt.Println("date=", Date)
		return nil
	}

	if scanComPorts {
		_, e := com.GetSerialPorts(w, verbose)
		return e
	}
	if hmp.SerialPortName == "" {
		fmt.Println(`no comport name, enter "hmp -help"`)
		return nil
	} else {
		if verbose {
			fmt.Println(`CLI switch '-p' exists, try to connect to HMP...`)
		}

		hmp.Power = hmp.NewDevice(w, hmp.SerialPortName, hmp.BaudRate, hmp.DataBits, hmp.Parity, hmp.StopBits, verbose)
		defer hmp.Power.Close()
		e := hmp.Power.Connect()
		if e != nil {
			log.Fatal(e)
		}
	}

	// example:

	hmp.Power.OutputOFF(-1)
	hmp.Power.SetVoltage(channel, voltage)
	hmp.Power.SetCurrent(channel, current)
	hmp.Power.OutputON(-1)
	for {
		hmp.Power.OutputON(channel)
		time.Sleep(time.Duration(timeOn) * time.Millisecond)
		if verbose {
			fmt.Print("channel:", channel, "V=", hmp.Power.Voltage(channel))
			fmt.Print("channel:", channel, "mA=", hmp.Power.Current(channel))
		}
		hmp.Power.OutputOFF(channel)
		time.Sleep(time.Duration(timeOff) * time.Millisecond)
		hmp.Power.OutputON(channel)
	}
}
